src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.DivRem.cs

```diff
@@ -79,44 +90,119 @@ public static void Divide(ReadOnlySpan<uint> left, ReadOnlySpan<uint> right, Spa
-                                  stackalloc uint[StackAllocThreshold]
-                                  : leftCopyFromPool = ArrayPool<uint>.Shared.Rent(left.Length)).Slice(0, left.Length);
-            left.CopyTo(leftCopy);
+                // NOTE: left will get overwritten, we need a local copy
+                // However, mutated left is not used afterwards, so use array pooling or stack alloc
+                Span<uint> leftCopy = (left.Length <= StackAllocThreshold ?
+                                      stackalloc uint[StackAllocThreshold]
+                                      : leftCopyFromPool = ArrayPool<uint>.Shared.Rent(left.Length)).Slice(0, left.Length);
+                left.CopyTo(leftCopy);
 
-            Divide(leftCopy, right, quotient);
+                DivideGrammarSchool(leftCopy, right, quotient);
+
+                if (leftCopyFromPool != null)
+                    ArrayPool<uint>.Shared.Return(leftCopyFromPool);
+            }
+            else
+                DivideBurnikelZiegler(left, right, quotient, default);
 
-            if (leftCopyFromPool != null)
-                ArrayPool<uint>.Shared.Return(leftCopyFromPool);
         }
 
         public static void Remainder(ReadOnlySpan<uint> left, ReadOnlySpan<uint> right, Span<uint> remainder)
         {
             Debug.Assert(left.Length >= 1);
             Debug.Assert(right.Length >= 1);
             Debug.Assert(left.Length >= right.Length);
-            Debug.Assert(remainder.Length >= left.Length);
+            Debug.Assert(remainder.Length == left.Length);
+            DummyForDebug(remainder);
 
-            // Same as above, but only returning the remainder.
+            if (right.Length < DivideBurnikelZieglerThreshold || left.Length - right.Length < DivideBurnikelZieglerThreshold)
+            {
+                // Same as above, but only returning the remainder.
 
-            left.CopyTo(remainder);
-            Divide(remainder, right, default);
+                left.CopyTo(remainder);
+                DivideGrammarSchool(remainder, right, default);
+            }
+            else
+            {
+                int quotientLength = left.Length - right.Length + 1;
+                uint[]? quotientFromPool = null;
+
+                Span<uint> quotient = (quotientLength <= StackAllocThreshold ?
+                                      stackalloc uint[StackAllocThreshold]
+                                      : quotientFromPool = ArrayPool<uint>.Shared.Rent(quotientLength)).Slice(0, quotientLength);
+
+                DivideBurnikelZiegler(left, right, quotient, remainder);
+
+                if (quotientFromPool != null)
+                    ArrayPool<uint>.Shared.Return(quotientFromPool);
+            }
         }
 
-        private static void Divide(Span<uint> left, ReadOnlySpan<uint> right, Span<uint> bits)
+        private static void DivRem(Span<uint> left, ReadOnlySpan<uint> right, Span<uint> quotient)
```

Comment:
nit: The fact that `left` is both the input and `remainder` is non-obvious. A small comment or renaming of the parameter would help the issue